crates/red_knot_python_semantic/src/types/call/bind.rs

```diff
@@ -0,0 +1,365 @@
+use super::{Argument, CallArguments, InferContext, Signature, Type};
+use crate::db::Db;
+use crate::types::diagnostic::{
+    INVALID_ARGUMENT_TYPE, MISSING_ARGUMENT, PARAMETER_ALREADY_ASSIGNED,
+    TOO_MANY_POSITIONAL_ARGUMENTS, UNKNOWN_ARGUMENT,
+};
+use crate::types::signatures::Parameter;
+use crate::types::UnionType;
+use ruff_python_ast as ast;
+
+/// Bind a [`CallArguments`] against a callable [`Signature`].
+///
+/// The returned [`CallBinding`] provides the return type of the call, the bound types for all
+/// parameters, and any errors resulting from binding the call.
+pub(crate) fn bind_call<'db>(
+    db: &'db dyn Db,
+    arguments: &CallArguments<'db>,
+    signature: &Signature<'db>,
+    callable_ty: Option<Type<'db>>,
+) -> CallBinding<'db> {
+    let param_count = signature.parameter_count();
+    let mut parameter_tys = vec![None; param_count];
+    let mut errors = vec![];
+    let mut next_positional = 0;
+    let mut first_excess_positional = None;
+    for (argument_index, argument) in arguments.iter().enumerate() {
+        let (index, parameter, argument_ty) = match argument {
+            Argument::Positional(ty) => {
+                let Some((index, parameter)) = signature
+                    .positional_at_index(next_positional)
+                    .map(|param| (next_positional, param))
+                    .or_else(|| signature.variadic_parameter())
+                else {
+                    first_excess_positional.get_or_insert(argument_index);
+                    next_positional += 1;
+                    continue;
+                };
+                next_positional += 1;
+                (index, parameter, ty)
+            }
+            Argument::Keyword { name, ty } => {
+                let Some((index, parameter)) = signature
+                    .keyword_by_name(name)
+                    .or_else(|| signature.keywords_parameter())
+                else {
+                    errors.push(CallBindingError::UnknownArgument {
+                        unknown_name: name.clone(),
+                        unknown_argument_index: argument_index,
```

Comment:
Nit: The `unknown` prefix feel a bit redundant. I'd just go with `argument_name` and `argument_index`